src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -855,6 +882,47 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 
 
 /*
+ * GetPlacementConnectionsReadyForTwoPhaseCommit iterates over the task placement list,
+ * starts the connections to the nodes and marks them critical. In the second iteration,
+ * the connection establishments are finished. Finally, BEGIN commands are sent,
+ * if necessary.
+ */
+static void
+GetPlacementConnectionsReadyForTwoPhaseCommit(List *taskPlacementList)
+{
+	ListCell *taskPlacementCell = NULL;
+	List *multiConnectionList = NIL;
+	ListCell *multiConnectionCell = NULL;
+
+	/* in the first iteration start the connections */
+	foreach(taskPlacementCell, taskPlacementList)
+	{
+		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+		int connectionFlags = SESSION_LIFESPAN;
+		MultiConnection *multiConnection = StartNodeConnection(connectionFlags,
+															   taskPlacement->nodeName,
+															   taskPlacement->nodePort);
+
+		MarkRemoteTransactionCritical(multiConnection);
+
+		multiConnectionList = lappend(multiConnectionList, multiConnection);
+	}
+
+	/* in the second iteration mark the connections as critical and finish the connection establishment */
+	foreach(multiConnectionCell, multiConnectionList)
```

Comment:
Hm. Let's add a helper finishing connection establishment for all connections in a list to connection_management.c? We can then make that more efficient later.